### DISCUSSION TITLE
How to "exit" or close gnuplot?

### DISCUSSION THREAD

**Discussion by jdnec** [alandefreitas/matplotplusplus#251] [Q&A] [answered]
My code has a loop that generates an image and takes a screenshot. 
But after the screenshot is taken, it doesn't exit gnuplot, and the gnuplot processes keeps accumulating consuming more and more RAM. 
These gnuplot processes only close after the loop is over. 
How do I exit or close the gnuplot process after taking the screenshot ? 
![Screenshot_1](https://user-images.githubusercontent.com/99940200/160914083-42e3205d-6021-4fae-9459-4b4735b25ece.jpg)
![Screenshot_2](https://user-images.githubusercontent.com/99940200/160914099-72812545-fb87-4963-9b20-1608e8ce89c4.jpg)

**Comment by alandefreitas** [comment]
If you call `show()` it should block until you press some key or you close gnuplot manually.

You can also just save the figure instead of manually taking a screenshot.

**Reply by jdnec** [reply]
> If you call `show()` it should block until you press some key or you close gnuplot manually.
> 
> You can also just save the figure instead of manually taking a screenshot.

I can't use show() because the loop stops, I need the loop to keep functioning without me pressing a key. 

Sorry when I said "take a screenshot", I meant, I saved the figure with matplot::save

I just need to save the picture and just afterwards close gnuplot while inside the loop.

**Reply by alandefreitas** [reply]
gnuplot should close when the figure is destroyed. Aren't you calling `figure` inside the loop?

**Reply by jdnec** [reply]
> gnuplot should close when the figure is destroyed. Aren't you calling `figure` inside the loop?

I don't understand. How can I destroy the figure?

Yes I'm calling a figure inside my loop, this how my code looks like (simplified):

for (int i=0; i<1000; i++)
{

                                        auto fig5 = matplot::figure(true);
                                        fig5->position({ 560, 1, 0, 0 });
                                        fig5->size(1920, 980);

                                        matplot::plot( x, y, x, z);
                                        matplot::save("temp/image.jpg");

                                       // destroy figure here?
}

Right now the figure is not being destroyed (or the gnuplot process doesn't end) .

**Comment by alandefreitas** [comment, marked as answer]
Don't call `auto fig5 = matplot::figure(true);` inside the loop. By creating a new figure, you're asking the backend to launch another instance of gnuplot to plot that chart. 

Just create the figure outside the loop and use the loop to update the figure:

```cpp
auto fig = matplot::figure(true);
for (int i=0; i<1000; i++)
{                                   
                                fig->position({ 560, 1, 0, 0 });
                                fig->size(1920, 980);
                                fig->plot( x, y, x, z);
                                fig->save("temp/image.jpg");
                               // figures destroyed with RAII when the handle gets invalidated
}
```

**Reply by jdnec** [reply]
> Don't call `auto fig5 = matplot::figure(true);` inside the loop. By creating a new figure, you're asking the backend to launch another instance of gnuplot to plot that chart.
> 
> Just create the figure outside the loop and use the loop to update the figure:
> 
> ```c++
> auto fig = matplot::figure(true);
> for (int i=0; i<1000; i++)
> {                                   
>                                 fig->position({ 560, 1, 0, 0 });
>                                 fig->size(1920, 980);
>                                 fig->plot( x, y, x, z);
>                                 fig->save("temp/image.jpg");
>                                // figures destroyed with RAII when the handle gets invalidated
> }
> ```

Oh, that's what was going on. Problem is solved now. 

Thanks!
